### PR TITLE
gh-96471: Add asyncio queue shutdown

### PR DESCRIPTION
**Pull request by EpicWink**
asyncio-only changes from #102499 (which supercedes #96474), updated to match the API introduced by #104750

* Issue: gh-96471

----
:books: Documentation preview :books:: https://cpython-previews--104228.org.readthedocs.build/

### CODE REVIEW

**Comment by EpicWink** [comment]
@willingc ready for review

**Comment by gvanrossum** [comment]
> asyncio-only changes from #102499 (on top of #96474)

I assume the "on top of ..." part is obsolete, since that PR was closed without merging. I'll add @willingc as a reviewer per your request.

**Comment by gvanrossum** [comment]
@EpicWink Could you merge main?

**Comment by EpicWink** [comment]
> Could you merge main?

Done, but like #104750, we may want to modify the implementation of `immediate=True` to simply consume the queue rather than having a separate state.

**Comment by gvanrossum** [comment]
Eh, @EpicWink, did you see the hubbub about the hanging test in your [previous PR](https://github.com/python/cpython/pull/104750)? The test was disabled because it kept hanging in CI on various platforms. Could you look into what's wrong with it? (At least the feature wasn't reverted.)

**Comment by EpicWink** [comment]
> Eh, @EpicWink, did you see the hubbub about the hanging test in your [previous PR](https://github.com/python/cpython/pull/104750)? The test was disabled because it kept hanging in CI on various platforms. Could you look into what's wrong with it?

I did see the follow-on issue and pull-requests. Yves and I are working on fixing it (you can follow the changes here: https://github.com/python/cpython/compare/main...EpicWink:cpython:fix-thread-queue-shutdown-test)

**Comment by gvanrossum** [comment]
Looking at this with an eye towards review, I realized that the threaded-queue implementation doesn't appear to match its documentation: It appears that `q.join()` does *not* raise `ShutDown` when the queue is shutdown immediately. The tests don't seem to be testing for this either. (Come to think of it, the test never seems to take the `except self.queue.ShutDown` path -- if I put a breakpoint there it never gets hit during *any* of the tests. The test logic is pretty convoluted, which is my only excuse for not having caught this in the review of gh-104750.)

**Comment by EpicWink** [comment]
Now that the immediate-consume implementation of the threading queue shutdown has been accepted, I'm going to do the same here and for multiprocessing. I'll personally rewrite the tests to be more readable and obvious.

Laurie
________________________________

File: Doc/library/asyncio-queue.rst
Comment (gvanrossum) [inline Doc/library/asyncio-queue.rst]:
```suggestion
      the queue is empty. Set *immediate* to true to make ``get`` raise
```

**Comment by gvanrossum** [comment]
> Now that the immediate-consume implementation of the threading queue shutdown has been accepted, I'm going to do the same here and for multiprocessing. I'll personally rewrite the tests to be more readable and obvious.

And what about `queue.Queue.join()`? Should it raise after `shutdown(immediate=True)`? Either way some changes are needed -- either in the implementation or in the docs. And the tests there too (what you wrote above *appears* to apply to the asyncio and multiprocessing versions only.)

**Review by gvanrossum — COMMENTED** [review]
I'll wait reviewing until you've rewritten shutdown, but here's one doc markup nit.

Also, please mark PRs that aren't ready for review as Draft.

**Comment by EpicWink** [comment]
> And what about `queue.Queue.join()`? Should it raise after `shutdown(immediate=True)`? Either way some changes are needed -- either in the implementation or in the docs. And the tests there too (what you wrote above _appears_ to apply to the asyncio and multiprocessing versions only.)

The behaviour should (and I intend to implement it to) be the same between the threading, multiprocessing and asyncio queues

**Comment by YvesDup** [comment]
By the way, should `task_done()` raise an exception after `shutdown` ?

**Comment by YvesDup** [comment]
Maybe I'm waking up a little too late, but despite having worked on the development 1 year ago, I have the impression that functionalities are missing in the merged version.

In the initial version, the basic shutdown just forbade `put` operations, also released all threads blocked in `put`. 
On the other hand, it left `get`, `join` and `task_done` operations possible.

The immediate shtdown prohibited all possible previous operations and had to release all threads blocked in `get`, and `join`. And the queue was purged.

Currently, in the following methods (`put`, `get`, `join` and `task`), there is no specific behavior depending on the type of stop. It's not even noted in an attribute.

I don't want to hurt anyone's feelings, but I think this PR needs to be reworked (including documentation).

**Comment by EpicWink** [comment]
> By the way, should `task_done()` raise an exception after `shutdown` or not ?

No (except the usual `ValueError`); see #115838

File: Doc/whatsnew/3.13.rst
Comment (gvanrossum) [inline Doc/whatsnew/3.13.rst]: I suspect it's `asyncio.Queue.shutdown`.

File: Lib/asyncio/queues.py
Comment (EpicWink) [inline Lib/asyncio/queues.py]:
```suggestion
        while self.empty():
            if self._is_shutdown and self.empty():
                raise QueueShutDown
```

I think I can reduce duplication with this and the next suggestion

File: Lib/asyncio/queues.py
Comment (EpicWink) [inline Lib/asyncio/queues.py]:
```suggestion
```

See previous comment

File: Lib/asyncio/queues.py
Comment (EpicWink) [inline Lib/asyncio/queues.py]:
```suggestion
        while self.full():
            if self._is_shutdown:
                raise QueueShutDown
```

I think I can reduce duplication with this and the next suggestion

File: Lib/asyncio/queues.py
Comment (EpicWink) [inline Lib/asyncio/queues.py]:
```suggestion
```

See previous comment

**Review by EpicWink — COMMENTED** [review]
I've removed @YvesDup's tests `test_shutdown_all_methods_in_one_task`, `test_shutdown_putters_deque`, `test_shutdown_getters_deque`, `test_shutdown_get`, `test_shutdown_get_task_done_join`, and `test_shutdown_put`, and replaced them with simpler and lower-level tests. Are there any scenarios I've missed?

**Comment by YvesDup** [comment]
> I've removed @YvesDup's tests `test_shutdown_all_methods_in_one_task`, `test_shutdown_putters_deque`, `test_shutdown_getters_deque`, `test_shutdown_get`, `test_shutdown_get_task_done_join`, and `test_shutdown_put`, and replaced them with simpler and lower-level tests. Are there any scenarios I've missed?

I agree with all the removes. 
May i suggest creating a test that checks  `_getters,  `_putters` and `_unfinished_tasks` private attributes when queue shutdowns, immediatly or not ?  Or adds them in the existing tests ?

File: Lib/test/test_asyncio/test_queues.py
Comment (YvesDup) [inline Lib/test/test_asyncio/test_queues.py]: `q.shutdown(immediate=False)` is more explicit. 
Same suggestion in line 562.

File: Lib/test/test_asyncio/test_queues.py
Comment (YvesDup) [inline Lib/test/test_asyncio/test_queues.py:672]: Could you please add a comment here about the queue state just after the shutdown ?
No item in the queue, self._unfinished_tasks is stiil set to 1.

File: Lib/test/test_asyncio/test_queues.py
Comment (YvesDup) [inline Lib/test/test_asyncio/test_queues.py]: In the `self._ping_awaitable` method, which block is executed for each of the 2 calls (lines 592 and 610)?  The `try` block or the `except` block.
Should we check?

Comment (EpicWink) [reply Lib/test/test_asyncio/test_queues.py]: The point of `_ping_awaitable` is to have the async loop run the code at all: it doesn't care how or what it raises or returns (that's checked elsewhere in the test, eg `assertTrue(join_task.done())` and `await join_task`)

File: Lib/test/test_asyncio/test_queues.py
Comment (EpicWink) [reply Lib/test/test_asyncio/test_queues.py:672]: I've added a comment indicating (expected) changes to unfinished-tasks, but for number of items I've instead added a check

File: Lib/test/test_asyncio/test_queues.py
Comment (YvesDup) [reply Lib/test/test_asyncio/test_queues.py]: Okay, so can I consider this to be equivalent to an `await asyncio.sleep(0)` instruction?
Or if I'm wrong, what's the difference?

**Comment by EpicWink** [comment]
> With your actual organization (`test_shutdown_empty`, `test_shutdown_nonempty`), its seems to me harder to follow the test purposes. It'is up to you.

Each test is a different scenario, and I'm checking (hopefully) the full externally-visible state and functionality for each scenario. I think I will add more comments in these tests.

File: Lib/test/test_asyncio/test_queues.py
Comment (EpicWink) [reply Lib/test/test_asyncio/test_queues.py]: `await asyncio.sleep(0)` will yield to the the next task scheduled, which after future code edits may not be `join_task`. This makes it explicit that `join_task` is depended on, and forces it to be started (if not already). I'll add this as a comment

**Comment by YvesDup** [comment]
> > With your actual organization (`test_shutdown_empty`, `test_shutdown_nonempty`), its seems to me harder to follow the test purposes. It'is up to you.
> 
> Each test is a different scenario, and I'm checking (hopefully) the full externally-visible state and functionality for each scenario. I think I will add more comments in these tests.

Thank for all the comments, this is really easier to understand. 
IMO an important use case is missing: the context is when there are an empty queue and a pending get. 
- when shutdown is not immediate, the task still stay in the `q._getters`and we have to cancel it.
- when shutdown is immediate, this task is removed in the `shutdown`method.

It could be based on following code:
```py
async def _test_shutdown_pending_get(immediate_=False):
    async def _get(q, res):
        try:
            r = await q.get()
            res.append(r)
        except asyncio.queues.QueueShutDown:
            res.append(True)
        except asyncio.CancelledError:
            res.append(False)

    res = []
    q = asyncio.Queue(1)
    loop = asyncio.get_running_loop()
    get_task = loop.create_task(_get(q, res))
    await _ensure_started(get_task)

    q.shutdown(immediate=immediate_)
    if not immediate_:
        get_task.cancel()

    await get_task
    assert(res[0] == immediate_)
```

**Comment by gvanrossum** [comment]
Hey guys, great progress here. Can you at-mention me on this PR when you're ready for review? I recommend merging main just before you do that (don't rebase please).

**Comment by EpicWink** [comment]
> IMO an important use case is missing: the context is when there are an empty queue and a pending get.
> 
>     * when shutdown is not immediate, the task still stay in the `q._getters`and we have to cancel it.
> 
>     * when shutdown is immediate, this task is removed in the `shutdown`method.

No, in both cases the task should raise `ShutDown`: when the queue is empty, there should be no distinction between immediate and non-immediate shutdown (ignoring unfinished-tasks). I'm fixing this bug and updating the test.

---

The failing multi-processing `WithProcessesTestBarrier` tests in the Windows 64-bit free-threading builds seem unrelated

**Comment by EpicWink** [comment]
@gvanrossum ready for review

**Comment by YvesDup** [comment]
> in both cases the task should raise `ShutDown`: when the queue is empty, there should be no distinction between immediate and non-immediate shutdown (ignoring unfinished-tasks).

@EpicWink , there is the same bug on threading queue shutdown. I will open an issue.

File: Lib/asyncio/queues.py
Comment (gvanrossum) [reply Lib/asyncio/queues.py]: This didn't happen -- are you still planning this change?

File: Lib/asyncio/queues.py
Comment (gvanrossum) [reply Lib/asyncio/queues.py]: Ditto.

File: Lib/asyncio/queues.py
Comment (gvanrossum) [inline Lib/asyncio/queues.py]: Nit: I like the "summary" part of a docstring to fit on one line.
```suggestion
        """Shut-down the queue, making queue gets and puts raise QueueShutDown.
```

File: Lib/asyncio/queues.py
Comment (gvanrossum) [inline Lib/asyncio/queues.py]: Not that it makes one iota of difference, but comparing this to the code in `task_done()` this could be `==`?
```suggestion
            if self._unfinished_tasks == 0:
```

**Review by gvanrossum — COMMENTED** [review]
Hi there! I'm torn. This looks great apart from 1-2 teensy nits, but the tests are a lot of code that I haven't fully grokked yet. I want to trust that you two have done the right thing there, but at the same time I remember that the threaded queue tests had some issue that caused random CI flakiness. While I ponder this I'm going to start a buildbots run to hopefully smoke out any test flakes.

Meanwhile, can you address my nits and ~~merge main (there's a simple conflict in what's new)~~, but not push until the big test run completes? (Buildbot tests can take many hours, sorry.)

**EDIT:** I went ahead and merged the conflict, since tests won't run when there's a conflict.

**Comment by gvanrossum** [comment]
ALso, why is the DO-NOT-MERGE label set? Is it because of [this](https://github.com/python/cpython/pull/104228/#issuecomment-2035201238)?

File: Lib/test/test_asyncio/test_queues.py
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_queues.py]: That this exists feels like a red flag. Also:
- Why is it a static method? It doesn't need self, but you always call it as `self._ensure_started(t)`. I wouldn't bother with the decorator.
- There's a 10 msec timeout, which is then ignored. What if it times out? Is there any kind of invariant this is supposed to ensure?
- If all you want to do is make sure that the task gets a chance to start, why not just `await asyncio.sleep(0)`? That's the standard idiom -- it starts/resumes all runnable tasks and runs all pending callbacks.

In conclusion, I recommend replacing all `await self._ensure_started(foo_task)` lines below with just `await asyncio.sleep(0)`.

File: Lib/test/test_asyncio/test_queues.py
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_queues.py]: Hm... I'm not a big fan of this kind of "poke internals" testing. Why not create a queue and immediately call `shutdown()` on it?

File: Lib/test/test_asyncio/test_queues.py
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_queues.py]: This will start all the tasks. In particular, `q.join()` will start and immediately complete (since the queue is empty at this point).

File: Lib/test/test_asyncio/test_queues.py
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_queues.py:568]: Since you have these I'm not sure why you *also* have a task exercising `q.get()`?

File: Lib/test/test_asyncio/test_queues.py
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_queues.py]: I think you could move this block up to just below the (non-failing) `q.task_done()` call.

File: Lib/test/test_asyncio/test_queues.py
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_queues.py:577]: Why not just `await q.put("data")`?

**Review by gvanrossum — COMMENTED** [review]
Hm... Now that I've actually looked, I think these tests work, but could be simplified a bunch. Here are my additional comments.

File: Lib/asyncio/queues.py
Comment (willingc) [reply Lib/asyncio/queues.py]: Yeah, I think `==` makes more sense since a negative value for `_unfinished_tasks` would signal something went wrong.

Comment (EpicWink) [reply Lib/asyncio/queues.py]: This is a hold-over from a hack-fix in the [multi-threaded implementation](https://github.com/python/cpython/pull/104750#discussion_r1483950194), and is not needed anymore

File: Lib/test/test_asyncio/test_queues.py
Comment (EpicWink) [reply Lib/test/test_asyncio/test_queues.py]: > That this exists feels like a red flag

I am depending on a specific order of operations in the test (other arrangements  are covered by other tests or are trivial)

---

> Why is it a static method?

Minor style preference of mine, I'll remove it to make it consistent with the repo

---

> There's a 10 msec timeout, which is then ignored.

I didn't realise I could use a timeout of 0

> What if it times out?

I'm testing the status and result of the tasks later in the tests, so I'm ignoring it here

---

> why not just `await asyncio.sleep(0)`?

I didn't know `await asyncio.sleep(0)` would allow all other tasks to run (I thought which task gets run next was random), and instead had an explicit version. I'll do the replace you suggest and test it

File: Lib/test/test_asyncio/test_queues.py
Comment (EpicWink) [reply Lib/test/test_asyncio/test_queues.py:568]: I'm testing `await q.get()` raises after shutting down here, but for the previous task the get() is started before the shut down

File: Lib/test/test_asyncio/test_queues.py
Comment (EpicWink) [reply Lib/test/test_asyncio/test_queues.py:577]: In this scenario the two are equivalent (the queue is not empty, there are no other tasks scheduled), so I thought `put_nowait` was more explicit

File: Lib/test/test_asyncio/test_queues.py
Comment (EpicWink) [reply Lib/test/test_asyncio/test_queues.py]: I don't understand why this makes any difference

File: Lib/test/test_asyncio/test_queues.py
Comment (EpicWink) [reply Lib/test/test_asyncio/test_queues.py]: I see, I didn't know `await asyncio.sleep(0)` yields to all scheduled tasks

**Comment by EpicWink** [comment]
> I want to trust that you two have done the right thing there

@gvanrossum please don't trust, have you seen `xz-utils`'s backdoor?

> but not push until the big test run completes? (Buildbot tests can take many hours, sorry.)

Oops, sorry. I forgot to hold off my most recent push

> ALso, why is the DO-NOT-MERGE label set? Is it because of https://github.com/python/cpython/pull/104228/#issuecomment-2035201238?

No, I think it was because I hadn't updated the implementation to match the new agreed-upon behaviour (this PR was in draft at that time anyway, so I don't think the label was needed).

> wondering why you always wake up all getters -- and came to the conclusion that it is needed

I should probably add a comment

**Comment by EpicWink** [comment]
> there is the same bug on threading queue shutdown. I will open an issue.

@YvesDup I've created issue #117531, and fix in PR #117532

File: Lib/test/test_asyncio/test_queues.py
Comment (YvesDup) [inline Lib/test/test_asyncio/test_queues.py:668]: I suggest:
```py
q.put_nowait("data1")
q.put_nowait("data2")
```

File: Lib/test/test_asyncio/test_queues.py
Comment (YvesDup) [inline Lib/test/test_asyncio/test_queues.py:671]: and testing on  `"data1"`.

**Comment by YvesDup** [comment]
> > there is the same bug on threading queue shutdown. I will open an issue.
> 
> @YvesDup I've created issue #117531, and fix in PR #117532

I have seen and commented

File: Lib/asyncio/queues.py
Comment (YvesDup) [reply Lib/asyncio/queues.py]: If you run that example:
```py
async def _test_shutdown_notempty_get():
    res = []
    q = asyncio.Queue()
    await q.put("xx")
    q.shutdown(False)
    await q.get() # Succeed !!!
```
No `ShutDown` exception raises, that's a problem because queue is shutdown.
See my comment about the `get_nowait` method.

File: Lib/asyncio/queues.py
Comment (YvesDup) [inline Lib/asyncio/queues.py:211]: You have to check if queue is shutdown before testing the empty state.
This should be:
```
    if self._is_shutdown:
        raise QueueShutDown
    if self.empty():
        raise QueueEmpty
```
Update: Tests are correct in the `put_nowait` method.

Comment (EpicWink) [reply Lib/asyncio/queues.py:211]: No, while the queue still has items, `get()` should always return those items, regardless of the value of `_shutdown`. That is the whole point of `immediate=False`.

File: Lib/asyncio/queues.py
Comment (EpicWink) [reply Lib/asyncio/queues.py]: No, that code example that you just posted is correct, and is expected to succeed. `shutdown(immediate=False)` is supposed to let consumers empty the queue before excepting

File: Lib/test/test_asyncio/test_queues.py
Comment (EpicWink) [reply Lib/test/test_asyncio/test_queues.py:671]: Except when `q` is an instance of `LifoQueue`, when the value should be `"data2"`. Is it necessary to test queue ordering in the shutdown tests?

File: Lib/asyncio/queues.py
Comment (YvesDup) [reply Lib/asyncio/queues.py:211]: > That is the whole point of immediate=False

You are right, inappropried remark, sorry !?

File: Lib/test/test_asyncio/test_queues.py
Comment (YvesDup) [reply Lib/test/test_asyncio/test_queues.py:671]: I forgot the `LifoQueue` class. Is this worth commenting on?

Comment (EpicWink) [reply Lib/test/test_asyncio/test_queues.py:671]: I have implemented your suggestion, but I have not committed. I am considering it. I don't think it's necessary.

Comment (YvesDup) [reply Lib/test/test_asyncio/test_queues.py:671]: The test case (thus including `LifoQueue`) is implicit. A comment would be enough.

**Review by gvanrossum — APPROVED** [review]
Yup! This version looks great. Let me know if you agree -- I've removed the DO-NOT-MERGE label, but I'll wait until you and @YvesDup are happy too.

**Comment by EpicWink** [comment]
I think it's good to go

<details><summary>.</summary>

> > wondering why you always wake up all getters -- and came to the conclusion that it is needed

> I should probably add a comment

I could do this, but I don't think it's worth revoking the PR's approval

</details>

File: Doc/library/asyncio-queue.rst
Comment (YvesDup) [inline Doc/library/asyncio-queue.rst:112]: Sorry but I have a doubt, shouldn't this documentation block be rather:
```     
    All blocked callers of :meth:`~Queue.put` and :meth:`~Queue.get` 
    will be unblocked. If *immediate* is true, also unblock callers of 
    :meth:`~Queue.join`.
```
In event of change,  the docstring of the `shutdown` method must be updated.

File: Lib/asyncio/queues.py
Comment (YvesDup) [inline Lib/asyncio/queues.py:261]: Docstring to modify depending of agree/disagree on my first remark about `blocked callers'.
Changes could be:
```
All blocked callers of put() and get() will be unblocked, and
also join() if 'immediate
```

File: Doc/whatsnew/3.13.rst
Comment (YvesDup) [inline Doc/whatsnew/3.13.rst:302]: We've worked well together, I'd be nice if you'd mention me. 
:-)

**Comment by gvanrossum** [comment]
Sorry for merging prematurely. @EpicWink Can you prepare a follow-up PR? I will merge it as soon as @YvesDup approves it. (Or, alternatively, @YvesDup can make a PR and I will merge when @EpicWink approves.)

File: Doc/library/asyncio-queue.rst
Comment (EpicWink) [reply Doc/library/asyncio-queue.rst:112]: `join` callers aren't necessarily even unblocked anyway, if consumers are processing any items. I should probably say that a task is marked as done for each item in the queue if immediate shutdown.

Also, I think the threading queue docs are the same.

Comment (YvesDup) [reply Doc/library/asyncio-queue.rst:112]: > a task is marked as done for each item in the queue if immediate shutdown.

It's very precise, better.

> Also, I think the threading queue docs are the same.

Yes, I commented here so as not to forget (see https://github.com/python/cpython/pull/117532#issuecomment-2041150494). 

English is your native language, I think it'is best if you update documentations and docstrings.

__Update:__ but I can create the follow-up PR.

**Comment by EpicWink** [comment]
@gvanrossum see #117621

File: Doc/library/asyncio-queue.rst
Comment (EpicWink) [reply Doc/library/asyncio-queue.rst:112]: I've made a PR: #117621

**Comment by john-parton** [comment]
I'd like to add that the `shutdown()` method and task groups makes certain patterns really easy to express. I was struggling with how to properly make a generic async map function with a fixed number of workers (mostly as an exercise), but when I saw the new shutdown method, it came together really easily:

```python3
async def async_map[T, R](
    func: Callable[[T], Awaitable[R]],
    iterable: AsyncIterable[T],
    *,
    limit: int,
    maxsize: int = -1,
) -> AsyncIterator[R]:
    if maxsize < 0:
        maxsize = limit

    arguments_queue = Queue[T](maxsize=maxsize)
    results_queue = Queue[R](maxsize=maxsize)

    async def drain():
        async for argument in iterable:
            await arguments_queue.put(argument)
        arguments_queue.shutdown(immediate=False)

    async def worker():
        while True:
            try:
                argument = await arguments_queue.get()
            except QueueShutDown:
                break
            await results_queue.put(await func(argument))

    async def background():
        async with asyncio.TaskGroup() as background_task_group:
            background_task_group.create_task(drain())
            for _ in range(limit):
                background_task_group.create_task(worker())
        results_queue.shutdown(immediate=False)
 
    async with asyncio.TaskGroup() as task_group:
        task_group.create_task(background())

        while True:
            try:
                yield await results_queue.get()
            except QueueShutDown:
                break
 ```
 
I believe this has a common issue that a lot of async generators do, in that if you don't consume the entire generator, it will still continue processing and end up with a lot of futures never awaited.

**Pull request merged**
